src/src/com/microsoft/aad/adal/StorageHelper.java

```diff
@@ -305,70 +283,241 @@ public String decrypt(String value) throws GeneralSecurityException, IOException
+                return mSecretKeyFromAndroidKeyStore;
+            }
+        }
+        
+        throw new IllegalArgumentException("keyVersion = " + keyVersion);
+    }
 
-        // IV: Initialization vector that is needed to start CBC
-        byte[] iv = new byte[DATA_KEY_LENGTH];
-        mRandom.nextBytes(iv);
-        IvParameterSpec ivSpec = new IvParameterSpec(iv);
+    @TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR2)
+    private synchronized KeyPair generateKeyPairFromAndroidKeyStore() throws GeneralSecurityException, IOException {
+        final KeyStore keyStore = KeyStore.getInstance("AndroidKeyStore");
+        keyStore.load(null);
+        
+        Logger.v(TAG, "Generate KeyPair from AndroidKeyStore");
+        Calendar start = Calendar.getInstance();
+        Calendar end = Calendar.getInstance();
+        end.add(Calendar.YEAR, 100);
+
+        // self signed cert stored in AndroidKeyStore to asym. encrypt key
+        // to a file
+        final KeyPairGenerator generator = KeyPairGenerator.getInstance("RSA",
+                "AndroidKeyStore");
+        generator.initialize(getKeyPairGeneratorSpec(mContext, start.getTime(), end.getTime()));
+        try {
+            return generator.generateKeyPair();
+        } catch (final IllegalStateException exception) {
+            // There is an issue with AndroidKeyStore when attempting to generate keypair
+            // if user doesn't have pin/passphrase setup for their lock screen. 
+            // Issue 177459 : AndroidKeyStore KeyPairGenerator fails to generate 
+            // KeyPair after toggling lock type, even without setting the encryptionRequired 
+            // flag on the KeyPairGeneratorSpec. 
+            // https://code.google.com/p/android/issues/detail?id=177459
+            // The thrown exception in this case is: 
+            // java.lang.IllegalStateException: could not generate key in keystore
+            // To avoid app crashing, re-throw as checked exception
+            ClientAnalytics.logEvent(new AndroidKeyStoreFailureEvent(
+                    new InstrumentationPropertiesBuilder(exception)));
+            throw new KeyStoreException(exception);
+        }
+    }
 
-        // Set to encrypt mode
-        Cipher cipher = Cipher.getInstance(CIPHER_ALGORITHM);
-        Mac mac = Mac.getInstance(MAC_ALGORITHM);
-        cipher.init(Cipher.ENCRYPT_MODE, mKey, ivSpec);
+    /**
+     * Read KeyPair from AndroidKeyStore. 
+     */
+    private synchronized KeyPair readKeyPair() throws GeneralSecurityException, IOException {
+        if (doesKeyPairExist()) {
+            // Read key pair again
+            Logger.v(TAG, "Reading Key entry");
+            try {
+                KeyStore keyStore = KeyStore.getInstance("AndroidKeyStore");
+                keyStore.load(null);
+                final KeyStore.PrivateKeyEntry entry = (KeyStore.PrivateKeyEntry)keyStore.getEntry(
```

Comment:
Check to see if getEntry can call containsAlias, if it could call containAlias, we still need to catch the NullPointerException that could be thrown.